### PR TITLE
Custodial account verification

### PR DESCRIPTION
**Pull request by GordyD**
<del>[WIP] Do not review yet :)</del>

### CODE REVIEW

**Comment by krystophv** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;8R9oDnbQ&#x2F;21-signup-account-creation-blip-signup-flow-changes-required-for-mvp-clinic-account-creation-deployment"> [signup] [account creation] Blip signup flow changes required for MVP clinic account creation deployment</a>

**Comment by jebeck** [comment]
Hmm, just trying this out for the first time locally with an account I created with uploader account creation. After verification success, I got a pop-up "An error occurred while logging in."

screenshot 2016-06-29 16 33 51

Not sure why login is being triggered as part of the series of steps and clearly the issue is the lack of a provided e-mail.

File: app/components/inputgroup/inputgroup.js [hunk 1]
```diff
@@ -17,15 +17,19 @@
 var React = require('react');
 var _ = require('lodash');
 
+var DatePicker = require('../datepicker');
+
 // Input with label and validation error message
 var InputGroup = React.createClass({
   propTypes: {
     name: React.PropTypes.string,
```
Comment (jebeck) [inline app/components/inputgroup/inputgroup.js]: Should any/all of these props be `isRequired`? The original component may date from before we decided on enforcing `propTypes` as part of our style.

File: app/core/utils.js [hunk 2]
```diff
@@ -22,6 +22,16 @@ import nurseShark from 'tideline/plugins/nurseshark';
...
+/**
+ * Convenience function for capitalizing a string
+ * 
+ * @param  {String} str
+ * @return {String}
+ */
+utils.capitalize = (str) => {
+  return str.charAt(0).toUpperCase() + str.slice(1)
+}
+
```
Comment (jebeck) [inline app/core/utils.js]: Unit test please? There's already a file for these and it'll take like 2sec /test/unit/utils/utils.test.js

File: app/core/utils.js [hunk 3]
```diff
@@ -161,8 +171,15 @@ utils.getSignupKey = function(location) {
 utils.getSignupEmail = function (location) {
```
Comment (jebeck) [inline app/core/utils.js]: Ditto on unit test, looks like there weren't any yet for this fn.

File: app/pages/verificationwithpassword/verificationwithpassword.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    this.resetFormStateBeforeSubmit(formValues);
+
+    var validationErrors = this.validateFormValues(formValues);
+
+    if (!_.isEmpty(validationErrors)) {
+      return;
+    }
+
+    formValues = this.prepareFormValuesForSubmit(formValues);
+    this.props.onSubmit(this.props.api, this.props.signupKey, this.props.signupEmail, formValues.birthday, formValues.password);
```
Comment (jebeck) [inline app/pages/verificationwithpassword/verificationwithpassword.js]: Huh, so the `signupEmail` is being passed in (unless it didn't get down to the component, though there should be a warning in the console if so...). Not sure why the post-success login attempt is failing on no e-mail.

File: test/unit/pages/verificationwithpassword.test.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+  describe('render', function() {
+    it('should render with 9 warnings when no props provided', function () {
+      console.error = sinon.stub();
+      let props = {};
+      let elem = React.createElement(VerificationWithPassword, props);
+      let render = TestUtils.renderIntoDocument(elem);
+      expect(console.error.callCount).to.equal(8);
+    });
+
+    it('should render without wanrings when all required props provided', function () {
```
Comment (jebeck) [inline test/unit/pages/verificationwithpassword.test.js]: Sp. fix on 'wanrings' plz

File: test/unit/pages/verificationwithpassword.test.js [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+import { VerificationWithPassword, mapStateToProps } from '../../../app/pages/verificationwithpassword/verificationwithpassword';
+
+describe('VerificationWithPassword', () => {
+  it('should be a function', () => {
+    assert.isFunction(VerificationWithPassword);
+  });
+
+  describe('render', function() {
+    it('should render with 9 warnings when no props provided', function () {
```
Comment (jebeck) [inline test/unit/pages/verificationwithpassword.test.js]: Description says 9 but test says 8

File: test/unit/pages/verificationwithpassword.test.js [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+    it('should return object with notification and working populated', () => {
+      let mapped = mapStateToProps(state);
+
+      expect(mapped.notification).to.equal(state.blip.working.verifyingCustodial.notification);
+      expect(mapped.working).to.equal(state.blip.working.verifyingCustodial.inProgress);
+    });
+  });
+
+  describe('isFormDisabled', () => {
```
Comment (jebeck) [inline test/unit/pages/verificationwithpassword.test.js]: A TODO?

File: app/pages/verificationwithpassword/verificationwithpassword.js (hunk 4 above)
Comment (krystophv) [inline app/pages/verificationwithpassword/verificationwithpassword.js]: I couldn't reproduce the login error. Are you setting the `signupEmail` as a `GET` variable in your initial URL?

File: test/unit/pages/verificationwithpassword.test.js (hunk 7 above)
Comment (krystophv) [inline test/unit/pages/verificationwithpassword.test.js]: I guess Gordon didn't get done with this. I'll see if I can't fill it in reasonably.

File: app/pages/verificationwithpassword/verificationwithpassword.js (hunk 4 above)
Comment (jebeck) [inline app/pages/verificationwithpassword/verificationwithpassword.js]: Ah, no, I was just putting the `signupKey` in. Assuming the link in real e-mails contains the e-mail (which makes sense), should be fine then.

File: app/core/utils.js (hunk 2 above)
Comment (krystophv) [inline app/core/utils.js]: 👍

File: app/pages/verificationwithpassword/verificationwithpassword.js (hunk 4 above)
Comment (krystophv) [inline app/pages/verificationwithpassword/verificationwithpassword.js]: I was doing the same thing. I looked up the email template in `hydrophone` and it does contain the `signupEmail` key in the URL.

**Comment by jebeck** [comment]
Did you look into this comment at all: https://github.com/tidepool-org/blip/pull/314#discussion_r69053432? Would like to know your thoughts on that, but otherwise LGTM.

**Comment by krystophv** [comment]
I took some time to reverse engineer what we are actually always passing in to the `InputGroup` component (via the `SimpleForm` component) and found that `type` is the only prop that we actually _always_ pass in. That prop seems to make some sense to me to be required, since a lot of the internal rendering logic depends on it.

**Pull request merged**